sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala

```diff
@@ -134,3 +139,70 @@ object ParserUtils extends SparkParserUtils {
     sb.toString()
   }
 }
+
+object LabelUtils {
+  private val seenLabels: mutable.Set[String] = mutable.Set.empty
+
+  def init(): Unit = {
+    seenLabels.clear()
+  }
+
+  private def checkLabels(
+      beginLabelCtx: Option[BeginLabelContext],
+      endLabelCtx: Option[EndLabelContext]) : Unit = {
+    (beginLabelCtx, endLabelCtx) match {
+      case (Some(bl: BeginLabelContext), Some(el: EndLabelContext))
+        if bl.multipartIdentifier().getText.nonEmpty &&
+          bl.multipartIdentifier().getText.toLowerCase(Locale.ROOT) !=
+            el.multipartIdentifier().getText.toLowerCase(Locale.ROOT) =>
+        withOrigin(bl) {
+          seenLabels.clear()
+          throw SqlScriptingErrors.labelsMismatch(
+            CurrentOrigin.get,
+            bl.multipartIdentifier().getText,
+            el.multipartIdentifier().getText)
+        }
+      case (None, Some(el: EndLabelContext)) =>
```

Comment:
Do you also need to explicitly handle case (Some(bl: BeginLabelContext), None), or this can fall into the default case _ below?

Reply:
This can fall into default `case _` because if _BeginLabel_ is defined and _EndLabel_ is not, that is always ok (except if BeginLabel is duplicate of some parent label but we are not concerned with that part here).